src/compiler/program.ts

```diff
@@ -1520,7 +1527,297 @@ namespace ts {
             }
         }
 
+        /**
+         * ExtensionKind.SyntacticLint or ExtensionKind.SemanticLint only
+         */
+        function performLintPassOnFile(sourceFile: SourceFile, kind: ExtensionKind): Diagnostic[] | undefined {
+            const lints = extensionCache.getCompilerExtensions()[kind];
+            if (!lints || !lints.length) {
+                return;
+            }
+            type UniqueLint = {name: string, walker: LintWalker, accepted: boolean, errored: boolean};
+            const initializedLints: UniqueLint[] = [];
+            const diagnostics: Diagnostic[] = [];
+            let activeLint: UniqueLint;
+            let parent: Node | undefined = undefined;
+            const profileLevel = options.profileExtensions;
+            for (const {name, args, ctor} of lints) {
+                let walker: LintWalker;
+                if (kind === ExtensionKind.SemanticLint) {
+                    const checker = getTypeChecker();
+                    startExtensionProfile(profileLevel, name, "construct", host.trace);
+                    try {
+                        walker = new (ctor as SemanticLintProviderStatic)({ ts, checker, args, host, program });
+                    }
+                    catch (e) {
+                        diagnostics.push(createExtensionDiagnostic(name, `Lint construction failed: ${(e as Error).message}`, sourceFile, /*start*/undefined, /*length*/undefined, DiagnosticCategory.Error));
+                    }
+                    completeExtensionProfile(profileLevel, name, "construct", host.trace);
+                }
+                else if (kind === ExtensionKind.SyntacticLint) {
+                    startExtensionProfile(profileLevel, name, "construct", host.trace);
+                    try {
+                        walker = new (ctor as SyntacticLintProviderStatic)({ ts, args, host, program });
+                    }
+                    catch (e) {
+                        diagnostics.push(createExtensionDiagnostic(name, `Lint construction failed: ${(e as Error).message}`, /*sourceFile*/undefined, /*start*/undefined, /*length*/undefined, DiagnosticCategory.Error));
+                    }
+                    completeExtensionProfile(profileLevel, name, "construct", host.trace);
+                }
+                if (walker) initializedLints.push({ name, walker, accepted: true, errored: false });
+            }
+
+            let nodesVisited = 0;
+            visitNode(sourceFile);
+
+            return diagnostics;
+
+            function visitNode(node: Node) {
+                node.parent = parent;
+                nodesVisited++;
+                let oneAccepted = false;
+                const oldParent = parent;

```

Comment:
I wanted to capture it before the node is handed off to extensions, _just in case they misbehave and edit it_. TBH, following that, I should do `node.parent = oldParent` right before `parent = node` - to ensure that the Node's parent is still correct when a child rule traverses up to it.

Now, given that any changes could still be visible to any other rules run after your rule on the same node, maybe I have to ensure it's set correctly after each rule execution...